### PR TITLE
Move `ImpossiblePolicy` from error to warning

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

* Deprecated `TypeErrorKind::ImpossiblePolicy`
* Added `ValidationWarningKind::ImpossiblePolicy` ⬅️ this requires a minor version bump
* Moved the definition of `ValidationWarning` to `validation_results.rs`
* Updated unit tests ⬅️ this accounts for most of the diff

As discussed in #539, `ImpossiblePolicy` is different from our other errors because it does not signal a potential authorization-time error. Instead, it indicates a policy that will not apply for any valid request.

As we increase the precision of the Cedar typechecker, `ImpossiblePolicy` errors will occur in more cases (a breaking change), while other types of errors will occur in fewer cases (a non-breaking change). So to make potential changes easier going forward, we've decided to demote `ImpossiblePolicy` to a warning. We consider this a _non-breaking_ change because all policies that previously validated will continue to validate.

## Issue #, if available

#539 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by khieta** [comment]
Note: the corpus test failure is expected since the tests were generated when we expected "impossible policy" to be an error (and thus the test says that the policy should not validate)

File: cedar-policy-validator/src/rbac.rs
Comment (cdisselkoen) [inline cedar-policy-validator/src/rbac.rs:1161]: is "notes" the generic term for warnings+errors?

File: cedar-policy-validator/src/rbac.rs
Comment (cdisselkoen) [inline cedar-policy-validator/src/rbac.rs:1338]: to avoid this `.clone()`, we could have `assert_validate_policy_succeeds()` take a `&Validator` which could be reused across multiple `assert_validate` calls

File: cedar-policy-validator/src/typecheck.rs
Comment (cdisselkoen) [inline cedar-policy-validator/src/typecheck.rs:271]:
```suggestion
    /// list, although these will not impact the boolean return value.
```

File: cedar-policy-validator/src/typecheck.rs
Comment (cdisselkoen) [inline cedar-policy-validator/src/typecheck.rs:277]: this boolean return value used by anyone? or ignored by all callers?

should callers just check `type_errors.is_empty()` and `warnings.is_empty()` to be more explicit about whether they are checking for the absence of warnings or not?

File: cedar-policy-validator/src/typecheck/test_optional_attributes.rs
Comment (cdisselkoen) [inline cedar-policy-validator/src/typecheck/test_optional_attributes.rs:781]: Despite the comment, this actually does fail, right? if "fail" means that validation fails, rather than that the test fails

File: cedar-policy-validator/src/rbac.rs
Comment (john-h-kastner-aws) [inline cedar-policy-validator/src/rbac.rs:1153]: We can use `#[track_caller]` on these

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-validator/src/rbac.rs
Comment (john-h-kastner-aws) [inline cedar-policy-validator/src/rbac.rs:1158]: Should this also check `.1` for warnings? Not strictly necessary for validation success, but it should be the case for the tests

**Review by john-h-kastner-aws — APPROVED** [review]

File: cedar-policy-validator/src/rbac.rs
Comment (khieta) [reply cedar-policy-validator/src/rbac.rs:1161]: I believe "notes" is the historical way to refer to errors... I'll update the terminology.

File: cedar-policy-validator/src/typecheck.rs
Comment (khieta) [reply cedar-policy-validator/src/typecheck.rs:277]: Return value is only used in tests -- I was on the fence about whether to remove it.

File: cedar-policy-validator/src/typecheck/test_optional_attributes.rs
Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/typecheck/test_optional_attributes.rs:781]: Yes. We changed this so that we know the attributes of action entities precisely (there typically being none, but code exists to handle attribute if they exist).

File: cedar-policy-validator/src/typecheck.rs
Comment (khieta) [reply cedar-policy-validator/src/typecheck.rs:277]: Decided to remove in the latest commit.

Comment (khieta) [reply cedar-policy-validator/src/typecheck.rs:277]: Scratch that -- turns out that the boolean value does matter sometimes. Will update comments accordingly.

Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/typecheck.rs:277]: We can still remove it if we want, but it would be beyond the scope of this PR

**Review by john-h-kastner-aws — APPROVED** [review]

**Comment by khieta** [comment]
Fyi: this PR will be on hold temporarily since I'll be OOTO until next Wednesday. I'm holding off on merging for now because I don't want to leave the corpus tests in a broken state.

**Comment by khieta** [comment]
I've pulled in the latest changes from `main`, so this PR is ready to merge (pending re-approval). Other PRs should expect the `Run cargo test --verbose --features "integration-testing" -- --ignored` step of CI to fail until I can manually update the corpus tests. I expect to do this either this weekend or early next week.

**Review by john-h-kastner-aws — APPROVED** [review]

**Pull request merged**
